Add redundancy, remove `https` option, add `onlyHttps` option…

ghostnumber7 opened the pull request:
Some things that could be considered to be useful:
* Multiple DNS resolver host alternatives
* Multiple DNS resolver providers
* Fallback to HTTPS by default if DNS does not resolve (no need to choose one or the other, just try DNS and if it fails use HTTPS)
* Ability to disable or enable HTTPS check (`{ https: true }` will only check HTTPS). Defaults to DNS + HTTPS
* Added `options.urls` array to be able to add more urls to fallback to (like `ifconfig.co`)

sindresorhus commented:
> DNS resolver hosts by host name not IP

Why? The IP address will not change anyway. Resolving to the IP adds a slight overhead.

> Multiple DNS resolver host alternatives
Multiple DNS resolver providers

Good idea. This should be documented in the readme.

> Fallback to HTTPS by default if DNS does not resolve (no need to choose one or the other, just try DNS and if it fails use HTTPS)

👍 

I don't really see the needs for a `dns` option though. I'm thinking the default behavior would be DNS + HTTPS fallback, and an option to force only HTTPS.

> Added options.urls array to be able to add more urls to fallback to (like ifconfig.co)

👍 

---

These changes will need tests and readme updates.

sindresorhus commented on index.js:
```diff
@@ -42,21 +80,57 @@ const queryDns = (version, options) => {
+	const promise = (async () => {
+		let internalPromise = Promise.resolve();
+
+		data.dnsServers.forEach(dnsServerInfo => {
```
I prefer `for-of` loops instead of forEach.

ghostnumber7 replied:
tried to use for-in but linter had problems with that and tried to do something that did not need ignoring it ... can change it back if wanted

sindresorhus commented on index.js:
```diff
@@ -42,21 +80,57 @@ const queryDns = (version, options) => {
+		data.dnsServers.forEach(dnsServerInfo => {
+			const {servers, question} = dnsServerInfo;
+			servers.forEach(server => {
+				// eslint-disable-next-line promise/prefer-await-to-then
```
The rule is there for a reason. There's never a case where you have to use `.then`.

ghostnumber7 replied:
I tried to do a for-in loop with await but linter was not happy with it ... in this case, I chain then after then so that they run concurrently to simulate a for loop with await inside ... maybe there's a better approach I'm not seeing

sindresorhus commented on index.js:
```diff
@@ -66,32 +140,53 @@ const queryDns = (version, options) => {
+							throw error;
+						}
+
+						return null;
```
I prefer using `undefined` and not `null`.

ghostnumber7 replied:
sure no problem

ghostnumber7 commented:
> > NS resolver hosts by host name not IP
> 
> Why? The IP address will not change anyway. Resolving to the IP adds a slight overhead.

Don't think overhead is relevant in this case. Thought about using hostnames when I realized that I couldn't find `2620:0:ccc::2` by digging any of the opendns resolvers. They may not change, but it made me doubt xD

ghostnumber7 commented:
I'm also thinking about randomizing the order in which dns hosts are checked to minimize the change of going multiple times to one that's having problems before going to others ... what do you think?

sindresorhus commented:
> I don't really see the needs for a dns option though. I'm thinking the default behavior would be DNS + HTTPS fallback, and an option to force only HTTPS.

⬆️

sindresorhus commented:
> Using ips instead of hostnames for DNS resolvers (probably :P)

I don't see any downside with it.

> Randomizing DNS resolve order?

👍 Good idea!

> Making changes on browser file to behave the same way?

👍

ghostnumber7 commented on readme.md:
```diff
@@ -41,9 +41,18 @@ Type: `object`
+Type: `array`<br>
+Default: `[]`
+
+In case you want to add your own custom HTTPS endpoints to get public IP from (like [ifconfig.co](https://ifconfig.co), por example), you can set them here. They will only be used if everything else fails.
```
sorry can't find the typo xD

ghostnumber7 replied:
never mind, found it :)

ghostnumber7 commented:
So ... did your suggestions, mocked in a less verbose way (don't know if I can make it better, I normally use jest instead of ava+sinon so to be honest, I have no idea what I'm doing XD). Removed the `-s` option on ava CLI call, but there is no option in package.json ava config to do the same, so had to add `serial` to every test. Also, ava is bad at handling async stubbing, so we either have every test on its own file, or we run serially (https://github.com/avajs/ava/issues/1066)
Did changes on browser file also.
Did find one downside to using ipv6 ip instead of hostnames for dns resolving: When you don't have ipv6 it would normally just fail to find your ipv6 ip when resolving via host name, but when using ipv6 ip, it times out instead, and is it's making 6 checks, it takes 30 seconds for that to happen on default config. I guess it would not be a problem because you will be using it when you know you have an ipv6 enabled network.
DNS option has been removed in previous commit, I updated the first comment now to reflect that.
I didn't do the randomizing on the DNS order ... maybe we can add a TODO comment somewhere and have it done at some other time :P

sindresorhus commented:
> but there is no option in package.json ava config to do the same

There is. See the `serial` setting. You can also do it for a single file by using the import `import {serial as test} from 'ava';`

> Also, ava is bad at handling async stubbing, so we either have every test on its own file, or we run serially (avajs/ava#1066)

The problem is neither AVA nor async, but rather concurrency. Stubbing is incompatible with concurrency.

> I didn't do the randomizing on the DNS order ... maybe we can add a TODO comment somewhere and have it done at some other time :P

Can you open an issue about it? TODO comments are code rot.

sindresorhus commented on index.js:
```diff
@@ -2,32 +2,78 @@
+			],
+			name: 'o-o.myaddr.l.google.com',
+			type: 'TXT',
+			clean: ip => ip.replace(/"/g, '')
```
I would call it `transform`

ghostnumber7 replied:
changed in favor of `transform`

sindresorhus commented on index.js:
```diff
@@ -42,21 +88,40 @@ const queryDns = (version, options) => {
+						}
+					} = dnsResponse;
+
+					const response = typeof data === 'string' ? data.trim() : data.toString().trim();
```
```suggestion
					const response = (typeof data === 'string' ? data : data.toString()).trim();
```

ghostnumber7 replied:
fixed

sindresorhus commented on test.js:
```diff
@@ -1,24 +1,68 @@
+	t.true(gotStub.called() > 0);
+});
+
+test.serial('IPv4 DNS failure opendns fallbacks to google DNS', async t => {
```
```suggestion
test.serial('IPv4 DNS failure OpenDNS falls back to Google DNS', async t => {
```

ghostnumber7 replied:
fixed

sindresorhus commented on test.js:
```diff
@@ -1,24 +1,68 @@
+	t.is(gotStub.called(), 0);
+});
+
+test.serial('IPv4 DNS failure fallbacks to HTTPS', async t => {
```
```suggestion
test.serial('IPv4 DNS failure falls back to HTTPS', async t => {
```

ghostnumber7 replied:
fixed

sindresorhus commented on readme.md:
```diff
@@ -41,9 +41,18 @@ Type: `object`
-Use a HTTPS check using the [icanhazip.com](https://github.com/major/icanhaz) service instead of the DNS query. [ipify.org](https://www.ipify.org) is used as a fallback if `icanhazip.com` fails. This check is much more secure and tamper-proof, but also a lot slower. **This option is only available in the Node.js version**.
+Use a HTTPS check using the [icanhazip.com](https://github.com/major/icanhaz) service instead of the DNS query. [ipify.org](https://www.ipify.org) is used as a fallback if `icanhazip.com` fails. This check is much more secure and tamper-proof, but also a lot slower. **This option is only available in the Node.js version**. Default behaviour is to check aginst DNS before using HTTPS fallback, if s ...
+
+##### urls
```
Maybe we should call it `fallbackUrls` to make its purpose clear?

ghostnumber7 replied:
changed to use `fallbackUrls `

sindresorhus commented:
> > I don't really see the needs for a dns option though. I'm thinking the default behavior would be DNS + HTTPS fallback, and an option to force only HTTPS.

> ⬆️ 

This is not implemented. There should only be a boolean (no `null`) option. Maybe called `onlyHttps`.

ghostnumber7 commented:
> > > I don't really see the needs for a dns option though. I'm thinking the default behavior would be DNS + HTTPS fallback, and an option to force only HTTPS.
> 
> > ⬆️
> 
> This is not implemented. There should only be a boolean (no `null`) option. Maybe called `onlyHttps`.

changed to `onlyHttps` default `false`

sindresorhus commented on index.d.ts:
```diff
@@ -1,18 +1,25 @@
 		readonly timeout?: number;
+
+		/**
+		Use a HTTPS check using the [icanhazip.com](https://github.com/major/icanhaz) service instead of the DNS query. [ipify.org](https://www.ipify.org) is used as a fallback if `icanhazip.com` fails. This check is much more secure and tamper-proof, but also a lot slower. **This option is only available in the Node.js version**. Default behaviour is to check aginst DNS before using HTTPS fallback, if ...
```
This description is just copy-pasted from `onlyHttps`.

ghostnumber7 replied:
I'm sorry I'm not sure what you mean by this. I mean. I know it is, though it should be, and I'm not sure if you would like me to remove it, copy properly or write something dedicated.

sindresorhus replied:
Why should it be? The text makes no sense in its current form for this option. The text should be about what the `fallbackUrls` option does.

sindresorhus replied:
This option also needs to be documented in the readme. The readme and index.d.ts should be in sync.

ghostnumber7 replied:
my bad ... I see that now :)

sindresorhus commented on index.d.ts:
```diff
@@ -1,18 +1,25 @@
+		Use a HTTPS check using the [icanhazip.com](https://github.com/major/icanhaz) service instead of the DNS query. [ipify.org](https://www.ipify.org) is used as a fallback if `icanhazip.com` fails. This check is much more secure and tamper-proof, but also a lot slower. **This option is only available in the Node.js version**. Default behaviour is to check aginst DNS before using HTTPS fallback, if ...
 
-		@default false
+		@default null
```
This default is not correct.

sindresorhus commented on index.d.ts:
```diff
@@ -24,7 +31,7 @@ declare const publicIp: {
 	Get your public IP address - very fast!
 
-	In Node.js, it queries the DNS records of OpenDNS which has an entry with your IP address. In browsers, it uses the excellent [icanhaz](https://github.com/major/icanhaz) service through HTTPS.
+	In Node.js, it queries the DNS records of OpenDNS, Google DNS and HTTPS services to determine your IP address. In browsers, it uses the excellent [icanhaz](https://github.com/major/icanhaz) service through HTTPS.
```
You need to mention ipify.org

sindresorhus commented on index.d.ts:
```diff
@@ -44,7 +51,7 @@ declare const publicIp: {
 	Get your public IP address - very fast!
 
-	In Node.js, it queries the DNS records of OpenDNS which has an entry with your IP address. In browsers, it uses the excellent [icanhaz](https://github.com/major/icanhaz) service through HTTPS.
+	In Node.js, it queries the DNS records of OpenDNS, Google DNS and HTTPS services to determine your IP address. In browsers, it uses the excellent [icanhaz](https://github.com/major/icanhaz) service through HTTPS.
```
You need to mention ipify.org

sindresorhus commented on readme.md:
```diff
@@ -2,7 +2,7 @@
 > Get your public IP address - very fast!
 
-In Node.js, it queries the DNS records of OpenDNS which has an entry with your IP address. In browsers, it uses the excellent [icanhaz](https://github.com/major/icanhaz) service through HTTPS.
+In Node.js, it queries the DNS records of OpenDNS, Google DNS and HTTPS services to determine your IP address. In browsers, it uses the excellent [icanhaz](https://github.com/major/icanhaz) service through HTTPS.
```
You need to mention ipify.org

sindresorhus commented:
Did you see https://github.com/sindresorhus/public-ip/issues/40? I think we should remove icanhazip.com. But that leaves us with only ipify.org for the browser. Do you know any other free IP service that works over HTTPS?

sindresorhus commented on test.js:
```diff
@@ -1,24 +1,61 @@
 import test from 'ava';
```
```suggestion
import {serial as test} from 'ava';
```

and then you can make it `test.serial` => `test`.

sindresorhus commented on mocks/stub.js:
```diff
@@ -0,0 +1,35 @@
+	const original = objectPath[propertyName];
+
+	function stub(...args) {
+		if (ignoreIndex !== null) {
```
```suggestion
		if (ignoreIndex !== undefined) {
```

sindresorhus commented on mocks/stub.js:
```diff
@@ -0,0 +1,35 @@
+const sinon = require('sinon');
+
+module.exports = (objectPath, propertyName, ignoreIndex = null) => {
```
```suggestion
module.exports = (objectPath, propertyName, ignoreIndex) => {
```

sindresorhus commented on browser.js:
```diff
@@ -6,13 +6,14 @@ const defaults = {
+	],
+	v6: [
+		'https://ipv6.icanhazip.com/',
+		'https://api6.ipify.org'
```
```suggestion
		'https://api6.ipify.org/'
```

sindresorhus commented on browser.js:
```diff
@@ -6,13 +6,14 @@ const defaults = {
-	v6: 'https://api6.ipify.org'
+	v4: [
+		'https://ipv4.icanhazip.com/',
+		'https://api.ipify.org'
```
```suggestion
		'https://api.ipify.org/'
```

sindresorhus commented on index.js:
```diff
@@ -2,32 +2,79 @@
 const dgram = require('dgram');
 const dns = require('dns-socket');
-const got = require('got');
+const {get: got, CancelError} = require('got');
```
```suggestion
const {got, CancelError} = require('got');
```

`get` is the default.

sindresorhus commented on readme.md:
```diff
@@ -38,12 +38,21 @@ Returns a `Promise<string>` with your public IPv4 or IPv6 address. Rejects on er
+Type: `string[]`<br>
+Default: `[]`
+
+In case you want to add your own custom HTTPS endpoints to get public IP from (like [ifconfig.co](https://ifconfig.co), for example), you can set them here. They will only be used if everything else fails.
```
You need to mention the requirement that the fallbacks here must return the IP as a plain string.

sindresorhus commented:
Very nice! Thanks for working on this 🙌
